### PR TITLE
Rename DUNE to DUNES

### PR DESCRIPTION
**Pull request by jolly-fellow**
Renames all names DUNE to DUNES corresponding to the issue: https://github.com/AntelopeIO/DUNE/issues/141

I'm not sure about names of the installation packages and names of directories and symlinks creating during the installation process, should they be changed, does it make sense to create the symlinks? 
And may be the tags and names in the GitHub and Docker repositories should be changed too?

### CODE REVIEW

**Review by mikelik — CHANGES_REQUESTED** [review]
I would rename all `dune` to `dunes`. Otherwise we will be confused where we should use which version.

Can you rename all instances of `antelopeio-dune` to `dunes`? `antelopeio-dune` was package name chosen because `dune` was taken.


What has left in each file:
```mikel@msi:~/repo/DUNE$ git ls-files | xargs grep -w dune
README.md:#### DUNE installation on Windows <a name="dune-windows"></a>
README.md:choco uninstall antelopeio-dune -y
README.md:This will start the `dune` container.
bootstrap.bat:docker build --no-cache -f Dockerfile.win -t dune %mypath%
bootstrap.sh:docker build --no-cache --build-arg USER_ID=0 --build-arg GROUP_ID="$GROUP_ID" $LEAP_ARGUMENT $CDT_ARGUMENT -f Dockerfile.unix -t dune "$SDIR"
docs/PLUGIN.md:   2.  (optionally) `set_dune(dune)` - function that receives instance of DUNES so the user could interact with DUNES. It might be stored for later usage if needed.
docs/PLUGIN.md:In this example you can see how `set_dune` function is being used to store `dune` instance and later use it to create and prepare accounts.
docs/PLUGIN.md:2. (optionally) `set_dune(dune)` - if the user wants to interact with DUNES, they should store the DUNES object in this function.
docs/RELEASE.md:docker tag dune ghcr.io/antelopeio/dune:latest
docs/RELEASE.md:docker push ghcr.io/antelopeio/dune:latest
docs/RELEASE.md:docker tag dune ghcr.io/antelopeio/dune:X.Y.Z
docs/RELEASE.md:docker push ghcr.io/antelopeio/dune:X.Y.Z
docs/man/make_man.sh:pandoc dune.1.md -s -t man -o dune.1
dunes:python3 $SDIR/src/dune "$@"
dunes.bat:  python %mypath%\src\dune %*
dunes.bat:    python3 %mypath%\src\dune %*
packaging/antelopeio-dune/antelopeio-dune.nuspec:    <id>antelopeio-dune</id>
packaging/generate_chocolatey.bat:set target-dir=%script-path%\antelopeio-dune\tools
packaging/generate_chocolatey.bat:copy "%script-path%"\..\dune* %target-dir%
packaging/generate_chocolatey.bat:cd %script-path%\antelopeio-dune
packaging/generate_chocolatey.bat:move %script-path%\antelopeio-dune\*.nupkg %mypath%
packaging/generate_deb.sh:mv "$PROJECT".deb "$NAME".deb  # rename DUNE.deb to antelopeio-dune<etc>.deb
packaging/generate_rpm.sh:#add PROJECT_PREFIX to avoid conflict with ocaml dune
scripts/write_context.sh:echo $1 > /app/.dune.ctx
src/dune/__main__.py:from dune import dune
src/dune/__main__.py:from dune import dune_error
src/dune/__main__.py:from dune import dune_node_not_found
src/dune/__main__.py:from dune import node
src/dune/__main__.py:from dune import version_full
src/dune/__main__.py:    dune_sys = dune(args)
src/dune/args.py:                    dune [ARGUMENTS] -- <COMMANDS> runs any number of commandline commands in the container.
src/dune/args.py:                    Example: dune -- cleos --help''')
src/dune/context.py:    _file_name = ".dune.ctx"
src/dune/docker.py:    _dune_url = 'ghcr.io/antelopeio/dune:latest'
src/dune/docker.py:                # download dune image
src/dune/docker.py:                    with subprocess.Popen(['docker', 'tag', self._dune_url, 'dune:latest']) as proc:
src/dune/docker.py:        self.execute_docker_cmd(['commit', 'dune', 'dune'])
src/dune/dune.py:class dune:
src/dune/dune.py:        self._docker = docker('dune_container', 'dune:latest', cl_args)
tests/common.py:DUNE_EXE = os.path.split(TEST_PATH)[0] + "/dune"
tests/container.py:    def __init__(self, container_name='dune_container', image_name='dune:latest'):
tests/container.py:        self.execute_docker_cmd(['commit', 'dune', 'dune'])
tests/show_untested_options.sh:# Find paths to the tests and the dune executable.
tests/show_untested_options.sh:DUNE="$DUNE_DIR/dune"
tests/test_container.py:    cntr = container('dune_container', 'dune:latest')
tests/test_deploy.py:    #   ./dune --debug --send-action myaccount hi ["test"] eosio@active
tests/test_help.py:    """Test that the output of `dune <some invalid option>` is as expected."""
tests/test_help.py:            b'usage: dune',
tests/test_help.py:            b'dune: error: unrecognized arguments: --some-invalid-option'
tests/test_help.py:    """Test that the output of `dune -h` and `dune --help` is as expected."""
tests/test_help.py:            b'usage: dune',
tests/test_keys.py:    cntr = container('dune_container', 'dune:latest')
tests/test_list_features.py:    """Test that the output of `dune --list-features` is as expected."""
tests/test_list_features.py:    # List of expected output lines from `dune --list-features`.
tests/test_monitor.py:    cntr = container('dune_container', 'dune:latest')
tests/test_nodes.py:    # Call dune, check the return is True.
tests/test_nodes.py:    """Validate the result of a call to `dune --simple-list` contains the
tests/test_nodes.py:    # Call dune, check the return is True.
tests/test_nodes.py:    """Validate the result of a call to `dune --simple-list` contains all
tests/test_nodes.py:    # Call dune, check the return is True.
tests/test_nodes.py:    # List of expected output lines from `dune --list`.
tests/test_nodes.py:    cntr = container('dune_container', 'dune:latest')
tests/test_nodes.py:    cntr = container('dune_container', 'dune:latest')
tests/test_project.py:    cntr = container('dune_container', 'dune:latest')
tests/test_remainder.py:    cntr = container('dune_container', 'dune:latest')
tests/test_version_selector.py:sys.path.insert(1, sys.path.append(dirname(realpath(__file__)) + "/../src/dune"))
tests/test_wallet.py:    cntr = container('dune_container', 'dune:latest')
```

**Comment by jolly-fellow** [comment]
> I would rename all `dune` to `dunes`. Otherwise we will be confused where we should use which version.

I agree but when we discussed this issue on the meeting and I asked this question about the code I got answer that the source code is out of the scope of this issue so I didn't change it.
 
> Can you rename all instances of `antelopeio-dune` to `dunes`? `antelopeio-dune` was package name chosen because `dune` was taken.

I can everything I just not sure that it should be changed everywhere. May be there is another agreement for names of the packages. In the package building scripts antelopeio-dune is creating from a prefix variable and a name variable. Where a prefix is "antelopeio" I was sure that it is an important part of the package name. If you are sure it may be removed and a package may be called dunes.deb I will remove it or you may remove it yourself. I have created this branch for the collaborative work.

**Review by ScottBailey — COMMENTED** [review]
I think we should revisit the issue of updating the source and files. A global search and replace is probably easier now than it will ever be.

For example, my first review step could be ensuring that both the following commands return empty:
```bash
find | rg -i dune | rg -iv dunes
rg -i dune | rg -iv dunes
```

I think doing it all at once, reviewing it all at once, and TESTING it all at once instead of drawing it out might be a better approach.

**Comment by jolly-fellow** [comment]
@mikelik I will finish soon the total renaming as you wish. Please check the packaging scripts please to make sure that I corrected the building of the package names right. And thanks for the list.

**Comment by jolly-fellow** [comment]
> I think we should revisit the issue of updating the source and files. A global search and replace is probably easier now than it will ever be.
> 
> For example, my first review step could be ensuring that both the following commands return empty:
> 
> ```shell
> find | rg -i dune | rg -iv dunes
> rg -i dune | rg -iv dunes
> ```
> 
> I think doing it all at once, reviewing it all at once, and TESTING it all at once instead of drawing it out might be a better approach.

I agree but I asked about the source code on the meeting and got explicit answer don't change it.
But well I will change the sourcecode in the last commit. It will be possible to roll it back.

**Comment by jolly-fellow** [comment]
By the way what we will do with name of the repository?

**Comment by jolly-fellow** [comment]
> > By the way what we will do with name of the repository?
> 
> Bucky will rename it just after we merge the PR.
> 
> Thanks for doing all the changes, I think it's the right approach.

So should I change the URLs now or after it will be changed?

**Comment by mikelik** [comment]
I would do it now and we can coordinate with Bucky

**Comment by jolly-fellow** [comment]
> I would do it now and we can coordinate with Bucky

And when official Docker image appearing in the repository? Its name should be changed too.

**Comment by mikelik** [comment]
> And when official Docker image appearing in the repository? Its name should be changed too.

Yes, this can be done just after merging as well.
New package `dunes` will have to be created here: https://github.com/orgs/AntelopeIO/packages/container/package/

**Comment by jolly-fellow** [comment]
Well I hope I have finished. Renamed all what I found, fixed couple of errors and reformatted some code to follow PEP8.

File: src/dunes/__main__.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def load_all_modules_from_dir(plugin_dir):
 
     handle_simple_args()
 
-    dune_sys = dune(args)
+    dune_sys = dunes(args)
```
Comment (ScottBailey) [inline src/dunes/__main__.py:83]: Consider renaming `dune_sys` to `dunes_sys`? 
```suggestion
    dunes_sys = dunes(args)
```

File: src/dunes/__main__.py [hunk 2]
```diff
@@ -80,7 +80,7 @@ def load_all_modules_from_dir(plugin_dir):
...
 
     for module in modules:
         if hasattr(module, 'set_dune'):
```
Comment (ScottBailey) [inline src/dunes/__main__.py:86]: I'm not sure what to do here, but a change is needed.

```python
    for module in modules:
        if hasattr(module, 'set_dunes'):
            module.set_dunes(dune_sys)
        elif hasattr(module, 'set_dune'):
            # prefer new `set_dunes`, so only test for `set_dune` if above does not exist.
            # probably need to warn the user this may not work? <-------------------- what to do here?
            module.set_dune(dune_sys)
```

File: src/dunes/__main__.py [hunk 3]
```diff
@@ -6,11 +6,11 @@
...
-from dune import node
-from dune import version_full
+from dunes import dunes
+from dunes import dunes_error
+from dunes import dunes_node_not_found
+from dunes import node
+from dunes import version_full
 
 
 def handle_version():
```
Comment (ScottBailey) [inline src/dunes/__main__.py:16]: handle version needs an update to:
`print("DUNES " + version_full())`

File: tests/common.py [hunk 4]
```diff
@@ -6,8 +6,8 @@
 TEST_PATH = os.path.dirname(os.path.abspath(__file__))
 
 # Set path for executable:
-DUNE_EXE = os.path.split(TEST_PATH)[0] + "/dune"
-print("Executable path: ", DUNE_EXE)
+DUNES_EXE = os.path.split(TEST_PATH)[0] + "/dune"
```
Comment (ScottBailey) [inline tests/common.py:9]: I'm pretty sure the tests can't pass without this change! :-) 
```suggestion
DUNES_EXE = os.path.split(TEST_PATH)[0] + "/dunes"
```

**Review by ScottBailey — CHANGES_REQUESTED** [review]
This looks great with just a couple really minor exceptions.

File: src/dunes/__main__.py (hunk 1 above)
Comment (jolly-fellow) [reply src/dunes/__main__.py:83]: renamed

File: src/dunes/__main__.py (hunk 3 above)
Comment (jolly-fellow) [reply src/dunes/__main__.py:16]: Thanks for your help. Renamed.

File: tests/common.py (hunk 4 above)
Comment (jolly-fellow) [reply tests/common.py:9]: Renamed.

File: packaging/generate_deb.sh
[the remark this answers is not in the document]
Comment (jolly-fellow) [reply packaging/generate_deb.sh:1]: > At minimum, we probably want to add the following between `Version` and `Depends` (approximately line 22).

Need for @mikelik opinion. I have no practical experience of building packages.

**Comment by jolly-fellow** [comment]
> This looks great with just a couple really minor exceptions.

Thanks for your help. I can't understand how I left these strings not renamed. :)

File: packaging/generate_deb.sh
[the remark this answers is not in the document]
Comment (mikelik) [reply packaging/generate_deb.sh:1]: @jolly-fellow mentioned on the standup that he is willing not to add anything to deb package

File: src/plugin/dunes_hello/main.py
Comment (ScottBailey) [inline src/plugin/dunes_hello/main.py:1]: I don't think this should have been added?

File: packaging/generate_deb.sh
[the remark this answers is not in the document]
Comment (ScottBailey) [reply packaging/generate_deb.sh:1]: @jolly-fellow  Adding `Replaces: ` is simple and straightforward. More importantly, I think it's "the right thing to do". I'm a bit concerned that we wouldn't want to do that. Could you provide some detail regarding your heitation?

[the remark this answers is not in the document]
Comment (jolly-fellow) [reply packaging/generate_deb.sh:1]: > 

Because I don't believe that there are users who use DUNE now for serious work and will have problems with reinstalling. IMHO people install it now for learning purposes only so adding of any back compatibility code now doesn't make sense.

File: src/dunes/__main__.py (hunk 2 above)
Comment (ScottBailey) [reply src/dunes/__main__.py]: > There is no any third party plugin so I just renamed to set_dunes as in the plugin examples. 

We might theorize there are no third party plugins, but do we really know? 

The following lines of code are inexpensive and would really help someone that's already added a plugin: 
```python
        # Warn users of outdated plugins.
	elif hasattr(module, 'set_dune'):
            print( f"WARNING: {module} uses old interface." )
```
I suppose the _probability_ of a 3rd party plugin is low enough to make this a nice to have bit of code that doesn't really add much overall. Feel free to either add the warning, or ignore and resolve this conversation as you like.

**Review by ScottBailey — CHANGES_REQUESTED** [review]
Please remove `plugins/dunes_hellp/main.py`. It seems like it was maybe accidentally added during testing. 

Tests pass for me now. :+1: 

Please address the packaging issue.

File: src/plugin/dunes_hello/main.py
Comment (jolly-fellow) [reply src/plugin/dunes_hello/main.py:1]: Removed

**Comment by jolly-fellow** [comment]
> Please remove `plugins/dunes_hellp/main.py`. It seems like it was maybe accidentally added during testing.
> 
> Tests pass for me now. +1
> 
> Please address the packaging issue.

removed

File: docs/man/dunes.1 [hunk 5]
```diff
@@ -1,166 +1,166 @@
...
-\f[B]\f[CB]\-\- <COMMANDS>\f[B]\f[R]
+\f[B]\f[CB]-- <COMMANDS>\f[B]\f[R]
 .IP
 .nf
 \f[C]
 Adding double dashes and space at the end of the command line appends trailing arguments to the current 
 command. If the current command is empty, trailing arguments are passed as though they are the command.
 
-Example: dune \-\- cleos \-\-help
+Example: dunes -- cleos --help
```
Comment (mikelik) [inline docs/man/dunes.1:26]: interesting, @ScottBailey: I thought backslash was needed here, but it seems it works as it is now (both `man` and `md`).

File: docs/RELEASE.md [hunk 6]
```diff
@@ -1,21 +1,21 @@
-# How to prepare new DUNE release?
+# How to prepare new DUNES release?
 
 ## How to generate a package on Windows?
-1. Edit `packaging\antelopeio-dune\antelopeio-dune.nuspec` and edit the current version in XML tag `version`
+1. Edit `packaging\antelopeio-dunes\dunes.nuspec` and edit the current version in XML tag `version`
```
Comment (mikelik) [inline docs/RELEASE.md:4]:
```suggestion
1. Edit `packaging\dunes\dunes.nuspec` and edit the current version in XML tag `version`
```

File: src/dunes/args.py [hunk 7]
```diff
@@ -136,7 +136,7 @@ def __init__(self):
                                        'and multisig contracts. If optional arguments are provided '
                                        'it creates specific CURRENCY (default "SYS") with maximum amount of '
                                        'MAX_VALUE and initial value of INITIAL_VALUE')
-        self._parser.add_argument('--send-action', nargs=4, action=fix_action_data,
```
Comment (mikelik) [inline src/dunes/args.py:139]: is this change on purpose? why `action` was removed?

File: docs/man/dunes.1 (hunk 5 above)
Comment (ScottBailey) [reply docs/man/dunes.1:26]: This is auto-generated with the tool - and it's different for different versions. Part of why I'm against committing files like this.

**Review by ScottBailey — APPROVED** [review]
Ready for merge to AntelopeIO
